Generalizing decomposition of parameterization spectrum

dante831 opened the pull request:
We deal with the issue #296 of decomposing parameterization contribution into two parts: kinetic energy and potential energy. Improvements are as follows: 

1. We add a more general form of this decomposition in documentation, `docs/equations/notation_layered.rst`. 
2. Changed code to implement this updated form in `pyqg/qg_model.py`, and added the same thing to `pyqg/layered_model.py`. 
3. Modified `tests/test_diagnostics.py` to reflect the above changes.

asross commented on pyqg/tests/test_diagnostics.py:
```diff
@@ -43,16 +43,8 @@ def test_paramspec_decomposition(rtol=1e-10):
-    paramspec_apeflux = np.real((matvec(m.a, apeflux_term) * dqh).sum(axis=0))
-    paramspec_keflux  = np.real((matvec(m.a,  keflux_term) * dqh).sum(axis=0))
+    paramspec_apeflux = m.get_diagnostic('paramspec_apeflux')
+    paramspec_keflux = m.get_diagnostic('paramspec_keflux')
```
Hmm, could we revert this change (or modify it so it's doing the computation a different way above)? We [test below](https://github.com/pyqg/pyqg/blob/8f819876c15b59eb11d28ab81bf16dd22feef50e/pyqg/tests/test_diagnostics.py#L51-L55) that these terms match the diagnostic values, and setting them to the exact same thing defeats the purpose of the test.

dante831 replied:
Ahh okay, I'll change to an explicit calculation. Thanks for pointing this out!

asross commented on pyqg/qg_model.py:
```diff
@@ -289,31 +289,39 @@ def _initialize_model_diagnostics(self):
+                self.S[0,1] = self.F1
+                self.S[1,0] = self.F2
+                self.S[1,1] = -self.F2
+        get_S_matrix(self)
```
(if we want to avoid spuriously recomputing this matrix we can also do 

```python
from functools import cached_property

@cached_property
def S(self):
   # ...
```
)

dante831 replied:
Good point... it's a bit unclean in my original code. I will create `self.S` in function `_initialize_inversion_matrix` in `qg_model.py` then.

dante831 commented on pyqg/layered_model.py:
```diff
@@ -362,4 +362,27 @@ def _initialize_model_diagnostics(self):
+        height_ratios = (self.Hi/self.H)[:,np.newaxis,np.newaxis]
+        return np.real(height_ratios*
+                (np.einsum("ij..., j... -> i...", self.a, term)*self._calc_parameterization_contribution())
+                ).sum(axis=0)
```
I initially thought this method does not work for surface QG model, since q in the surface-QG system is zero. But the current code should technically for that subclass. Moving to `Model` instead.

dante831 commented:
Updated code according to @asross 's comments. The split of parameterization contribution is now in the main file, with a check to make sure that the subclass models have matrix `S` implemented. 
I also added the calculation of enstrophy budget to `qg_model.py` (the same as those in `layered_model.py`). The main reason to just copy & paste instead of absorbing  into the main class `model.py` is that the enstrophy budget needs to calculate intermediate terms in `_calc_derived_fields`, which is implemented in subclasses. These intermediate does not make sense in other subclasses such as `sqg_model.py`.

dante831 commented:
Added units to all diagnostics that currently do not have units, and added `/self.M**2` to normalize diagnostics. I think it makes sense to refactor repetitive code and create a common structure, but maybe it should not happen in this pull request since it already has a lot happening. Opening issue #299 instead.

asross commented:
> Added units to all diagnostics that currently do not have units, and added `/self.M**2` to normalize diagnostics. I think it makes sense to refactor repetitive code and create a common structure, but maybe it should not happen in this pull request since it already has a lot happening! Opening issue #299 instead.

Sounds good to me! I'm happy to make that change as well. I think let's go ahead and merge this!!
